plugins/interactive/interactive.go

```diff
@@ -0,0 +1,287 @@
+		Properties: schematypes.Properties{
+			"disableDisplay": schematypes.Boolean{},
+			"disableShell":   schematypes.Boolean{},
+		},
+	}
+	if !p.config.ForbidCustomArtifactPrefix {
+		s.Properties["artifactPrefix"] = schematypes.String{
+			MetaData: schematypes.MetaData{
+				Title: "Artifact Prefix",
+				Description: "Prefix for the interactive artifacts will be used to " +
+					"create `<prefix>/shell.html`, `<prefix>/display.html` and " +
+					"`<prefix>/sockets.json`. The prefix defaults to `" +
+					p.config.ArtifactPrefix + "`",
+			},
+			Pattern:       `^[\x20-.0-\x7e][\x20-\x7e]*/$`,
+			MaximumLength: 255,
+		}
+	}
+	return schematypes.Object{
+		Properties: schematypes.Properties{
+			"interactive": s,
+		},
+	}
+}
+
+func (p *plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (
+	plugins.TaskPlugin, error,
+) {
+	var P payload
+	if schematypes.MustMap(p.PayloadSchema(), options.Payload, &P) != nil {
+		return nil, engines.ErrContractViolation
+	}
+	// If not always enabled or no options are given then this is disabled
+	if P.Interactive == nil && !p.config.AlwaysEnabled {
+		return nil, nil
+	}
+
+	// Extract options
+	o := opts{}
+	if P.Interactive != nil {
+		o = *P.Interactive
+	}
+	if o.ArtifactPrefix == "" || p.config.ForbidCustomArtifactPrefix {
+		o.ArtifactPrefix = p.config.ArtifactPrefix
+	}
+
+	return &taskPlugin{
+		opts:   o,
+		log:    options.Log,
+		parent: p,
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	parent           *plugin
+	log              *logrus.Entry
+	opts             opts
+	sandbox          engines.Sandbox
+	context          *runtime.TaskContext
+	shellURL         string
+	shellServer      *ShellServer
+	displaysURL      string
+	displaySocketURL string
+	displayServer    *displayServer
+}
+
+func (p *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	p.context = context
+	return nil
+}
+
+func (p *taskPlugin) Started(sandbox engines.Sandbox) error {
+	p.sandbox = sandbox
+
+	// Setup shell and display in parallel
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	var err1, err2 error
+	go func() {
+		err1 = p.setupShell()
+		wg.Done()
+	}()
+	go func() {
+		err2 = p.setupDisplay()
+		wg.Done()
+	}()
+	wg.Wait()
+
+	// Return any of the two errors
+	if err1 != nil {
+		return fmt.Errorf("Setting up interactive shell failed, error: %s", err1)
+	}
+	if err2 != nil {
+		return fmt.Errorf("Setting up interactive display failed, error: %s", err2)
+	}
+
+	err := p.createSocketsFile()
+	if err != nil {
+		return fmt.Errorf("Failed to create sockets.json file, error: %s", err)
+	}
+	return nil
+}
+
+func (p *taskPlugin) Stopped(_ engines.ResultSet) (bool, error) {
+	if p.shellServer != nil {
+		p.shellServer.Abort()
+	}
+	if p.displayServer != nil {
+		p.displayServer.Abort()
+	}
+	return true, nil
+}
+
+func (p *taskPlugin) Dispose() error {
+	if p.shellServer != nil {
+		p.shellServer.Abort()
+		p.shellServer.Wait()
+	}
+	if p.displayServer != nil {
+		p.displayServer.Abort()
+	}
```

Comment:
Done, I suspect it's overkill as stopping http servers is cheap.. .where as the setup does I/O uploading files...